% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/MiXcan_extract_summary.R,
%   R/MiXcan_refit_summary.R
\name{MiXcan_refit_summary}
\alias{MiXcan_refit_summary}
\title{Extract the gene expression prediction summary}
\usage{
MiXcan_refit_summary(model, x, y, pi, cov = NULL)

MiXcan_refit_summary(model, x, y, pi, cov = NULL)
}
\arguments{
\item{model}{A direct output from MiXcan() function, which includes the
prediction weights as well as other summaries of the prediction models.}

\item{y:}{The pre-cleaned expression level data for a single gene in N samples.}

\item{x:}{A N by P matrix for all the genetic predictors used to predict the genetically regulated expression  of the gene.}

\item{cov:}{A N by Q matrix for the covariates adjusted in the model (e.g. age, population stratification).}

\item{pi:}{An estimation of cell-type fraction for the cell type of interest.}
}
\value{
A data frame with weight for cell 1 and 2, including the potential meta data for the SNP

A data frame with weight for cell 1 and 2, including the potential meta data for the SNP
}
\description{
Extract the gene expression prediction summary (e.g. No of SNPs, R^2, correlation) function.
The output can be directly applied to GWAS data for cell-type specific TWAS.

Extract the gene expression prediction summary (e.g. No of SNPs, R^2, correlation) function.
The output can be directly applied to GWAS data for cell-type specific TWAS.
}
